### DISCUSSION TITLE
[Plugin] Enable built-in support for javascript include instructions

### DISCUSSION THREAD

**Discussion by frutiemax** [OpenRCT2/OpenRCT2#21984] [Feature requests]
It would be great that OpenRCT2 could recognize javascript include instructions so it would be easier to split the code of a plugin into multiple files.

Some things to consider:

- File names should be relative to the path in which the javascript file with the include instruction is used.
- There should be a protection against infinite/cyclic file inclusions.
- If the file is not found, throw a warning at the console, but don't stop the execution.

**Comment by oli414** [comment]
You can do this using a transpiler. The version of JavaScript used in OpenRCT2 doesn't support the "import" tag yet (since it is ES5), but I don't think it is up to the OpenRCT2 devs to implement this feature from ES6:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import

You can use an NPM package like Babel and rollup to transpile your multi-file ES6 project into a single ES5 file.

**Comment by mgovea** [comment]
I attempted to use "require" which is the ES5 equivalent, and it didn't work either.

**Comment by oli414** [comment]
@mgovea "require" isn't part of the ES5 spec, but rather something that Node.js uses.

I've created a boilerplate npm project that allows you to transpile a multi-file ES6 project to ES5 with as little work as possible:
https://github.com/oli414/openrct2-plugin-boilerplate
This is a very common workflow for JavaScript projects since ES6 is still considered very new.

**Comment by IntelOrca** [comment]
duktape does support `require`, but we must implement the routine to supply the data which we do not currently do.
